File: day12/day12b.c
#include <stdio.h>
#include <string.h>

int main() {
    int input[4][3] = {
        {-7, -1, 6},
        {6, -9, -9},
        {-12, 2, -7},
        {4, -17, -12}
    };

    /*
    int input[4][3] = {
        {-1, 0, 2},
        {2, -10, -7},
        {4, -8, 8},
        {3, 5, -1}
    };
    */

    int loops[3];

    for (int i=0; i<3; i++) { // coord
        int initial[4];
        for (int j=0; j<4; j++) initial[j] = input[j][i];
        int positions[4];
        memcpy(positions, initial, sizeof(positions));
        int velocities[4] = {0};
        for (int step=1; step<=1000000; step++) {
            int allzero = 1;
            for (int j=0; j<4; j++) { // planet
                int a = (positions[j] < positions[0]) - (positions[j] > positions[0]) +
                    (positions[j] < positions[1]) - (positions[j] > positions[1]) +
                    (positions[j] < positions[2]) - (positions[j] > positions[2]) +
                    (positions[j] < positions[3]) - (positions[j] > positions[3]);
                velocities[j] += a;
                if (velocities[j]) allzero = 0;
            }
            for (int i=0; i<4; i++) {
                positions[i] += velocities[i];
            }
            if (memcmp(positions, initial, sizeof(positions)) == 0 && allzero) {
                loops[i] = step;
                break;
            }
        }
    }

    printf("%d %d %d\n", loops[0], loops[1], loops[2]);
}

File: day12/day12b.nim
# every little thing I think I'll want to do requires a fucking import
# I thought the python docs were bad, but this is next level bad
# there are no examples, anywhere
# just stupid function signatures
# I DON'T WANT TO READ SIGNATURES UNLESS I HAVE A BUG
# SIGNATURES ARE FOR THE COMPILER TO READ
#import strformat
import strutils
import sequtils
import math

#let lines = split(input, '\n')
# errors on this language suck
var ipositions : array[4, array[3, int]]
var positions : array[4, array[3, int]]
var velocities : array[4, array[3, int]]
var loops : array[3, int]

var i = 0
# streams dont implement pairs() (enumerate)
for tainted_line in lines(stdin):
  let line = tainted_line.string.strip(chars = {'<', '>'})
  let parts = line.split(", ")
  # but arrays do
  for j, part in parts:
    let xv = part.split("=")
    positions[i][j] = parseInt(xv[1])
  i += 1

ipositions = positions

for i in 0..2: # coord
  for step in 1..10000000:
    for j in 0..3: # moon
      for k in 0..3: # other moon
        velocities[j][i] += (if positions[j][i] < positions[k][i]: 1 else: 0) -
                            (if positions[j][i] > positions[k][i]: 1 else: 0);
    var all_eq = true
    var all_zero = true
    for j in 0..3: # moon
      positions[j][i] += velocities[j][i]
      if positions[j][i] != ipositions[j][i]:
        all_eq = false
      if velocities[j][i] != 0:
        all_zero = false
    if all_eq and all_zero:
      #echo "yay", step, " ", positions, " ", ipositions
      loops[i] = step
      break

#echo lcm(loops[0], lcm(loops[1], loops[2]))
echo loops.foldl(lcm(a, b)) # magic a and b instead of foldl(lcm) what is this
